source/common/router/router.cc

```diff
@@ -818,21 +812,33 @@ void Filter::UpstreamRequest::decode100ContinueHeaders(Http::HeaderMapPtr&& head
 }
 
 void Filter::UpstreamRequest::decodeHeaders(Http::HeaderMapPtr&& headers, bool end_stream) {
+  // TODO(rodaine): This is actually measuring after the headers are parsed and not the first byte.
+  parent_.callbacks_->requestInfo().firstUpstreamRxByteReceived(std::chrono::steady_clock::now());
```

Comment:
Do we ever do anything other than pass `std::chrono::steady_clock::now()` to any of these functions? I wonder ic it would better to actually have request info snap the current time and make these functions more like `onFirstUpstreamRxByteReceived()`. The benefit of that also is you can ASSERT it is never called twice to populate the optional twice. I think this would be an improvement so throwing it out there. (Obviously this applies to all cases of this pattern).

Reply:
Yeah, other than tests it's exclusively `steady_clock::now()`. I'm cool with this change. As for the ASSERTs though, in the cases of retries, we do expect it to be called multiple times, no?

Reply:
When I originally commented on this I thought about retries, which is kind of an interesting case. I actually don't think the asserts should be an issue for retries. Here is why:

1) In the upstream log case, we should be getting discrete data points for each try.
2) In the downstream log case, the router should only ever being to forward (and time) a single request. If this is not currently the case, you should be able to rejigger the router code slightly to make it so.

I think adding ASSERTs is a good thing to do here. We can always remove them later if need be. You have been looking at this code a lot more recently so feel free to correct me if what I'm saying isn't true.